r/R/expression.R

```diff
@@ -269,40 +266,40 @@ Expression$scalar <- function(x) {
   dataset___expr__scalar(Scalar$create(x))
 }
 
-build_dataset_expression <- function(.Generic, e1, e2, ...) {
-  if (.Generic %in% names(.unary_function_map) || nargs() == 2L) {
-    expr <- Expression$create(.unary_function_map[[.Generic]] %||% .Generic, e1)
-  } else if (.Generic == "%in%") {
+build_dataset_expression <- function(FUN,
+                                     ...,
+                                     args = list(...),
+                                     options = empty_named_list()) {
+  if (FUN == "%in%") {
     # Special-case %in%, which is different from the Array function name
-    expr <- Expression$create("is_in", e1,
+    expr <- Expression$create("is_in", args[[1]],
       options = list(
-        value_set = Array$create(e2),
+        value_set = Array$create(args[[2]]),
         skip_nulls = TRUE
       )
     )
   } else {
-    if (!inherits(e1, "Expression")) {
-      e1 <- Expression$scalar(e1)
-    }
-    if (!inherits(e2, "Expression")) {
-      e2 <- Expression$scalar(e2)
-    }
+    args <- lapply(args, function(x) {
+      if (!inherits(x, "Expression")) {
+        x <- Expression$scalar(x)
+      }
+      x
+    })
 
     # In Arrow, "divide" is one function, which does integer division on
     # integer inputs and floating-point division on floats
-    if (.Generic == "/") {
+    if (FUN == "/") {
       # TODO: omg so many ways it's wrong to assume these types
-      e1 <- e1$cast(float64())
-      e2 <- e2$cast(float64())
-    } else if (.Generic == "%/%") {
+      args <- lapply(args, function(x) x$cast(float64()))
+    } else if (FUN == "%/%") {
       # In R, integer division works like floor(float division)
-      out <- build_dataset_expression("/", e1, e2)
+      out <- build_dataset_expression("/", args = args)
```

Comment:
Does any of this change what kinds of error messages happen where there are illegal numbers of operands (e.g. ```` `/`(1,2,3)````)?

Reply:
If I read the code correctly, no. In the dplyr context, the expression builders on master would pass `e1, e2, ...` on through, so either way you'll get a message from the C++ library saying something like "divide_checked takes 2 arguments but you gave 3".